### PR TITLE
Migrate to core24 and python 3.12

### PR DESCRIPTION
**Pull request by upils**
Migrate imagecraft to core24 to avoid relying on 2 cores (since ubuntu-image is also now using core24) and so use python3.12 in the snap.

Python3.12 is explicitly used for now because otherwise python3.10 is picked up during the build. This will need to be removed at some point.

Also add a first very basic spread test to:
- make sure we test the snap is building and running fine
- prepare for more tests

### CODE REVIEW

**Comment by upils** [comment]
Even though every CI check is green, the build of the snap is broken. This needs to be fixed before merging.

**Comment by upils** [comment]
This is blocked by https://github.com/canonical/craft-parts/issues/786

**Comment by upils** [comment]
@sil2100 2 tests are still failing because we do not have the runners setup yet. Sergio from the snapd team should set them up today or this week.

File: Makefile
Comment (sil2100) [inline Makefile:16]: I didn't even know that the clean state also required passing the provider to use...

File: pyproject.toml
Comment (sil2100) [inline pyproject.toml:17]: Are we sure 3.11 is the minimal, or should that be 3.12 here as well?

File: setup.cfg
Comment (sil2100) [inline setup.cfg:22]: Same question as above.

File: snap/snapcraft.yaml
Comment (sil2100) [inline snap/snapcraft.yaml:45]: I know we're still in testing, but is using edge here always a good idea? What's the concept here? Is it because the stable ubuntu-image isn't core24-base yet, so temporary?

**Review by sil2100 — DISMISSED** [review]
LGTM! Do we want to wait with merging it until Sergio is done? Also, left some inline comments, but all might be irrelevant (just questions).

File: Makefile
Comment (upils) [reply Makefile:16]: I didn't either until someone from the craft team asked if I did that. I have to say I did not checked if this is really making a difference but at least we are "sure" this is cleaning what we are using to build.

File: pyproject.toml
Comment (upils) [reply pyproject.toml:17]: I was also wondering. Because if we consider the tool is to be used only as a snap then we do not really care for previous python versions. I am personally in favor of dropping support for python 3.11 (less linter fights, less tests to run, etc.) but I may miss some contraints. What do you think about that?

File: snap/snapcraft.yaml
Comment (upils) [reply snap/snapcraft.yaml:45]: Sorry I will add a comment. This is "needed" because otherwise installing the current stable version of ubuntu-image pulls core22 and thus python3.10. And according to the craft team this might be the cause of the issue I had (python3.10 was picked up when building the `imagecraft` part instead of the theoretically default python3.12). I am really not sure about that so as soon as we land a core24 based ubuntu-image in latest/stable I will remove that because I do not want imagecraft to rely on the edge channel of ubuntu-image.

**Comment by upils** [comment]
> LGTM! Do we want to wait with merging it until Sergio is done? Also, left some inline comments, but all might be irrelevant (just questions).

Yes I would like to wait for the setup of the runners to make sure this is working fine. I copied the conf from ubuntu-image but I may have messed it up so I would like to be sure the whole thing will properly work before merging.

**Comment by upils** [comment]
@sil2100 I have added snapd-testing-tools as a subtree (like in ubuntu-image) so do not worry about the files added in `tests/lib/external/`. As previously mentioned, due to https://github.com/canonical/craft-parts/issues/786, we cannot really merge this sadly because otherwise the tool will only work on 24.04. 

With Sergio we are still trying to understand why one of the runner is not working properly but otherwise once we decide if we keep python3.11 or not, the code looks ready for a last review.

File: pyproject.toml
Comment (sil2100) [reply pyproject.toml:17]: I'd be +1 on that. Doesn't have to be changed as part of this PR though.

**Review by sil2100 — APPROVED** [review]
I don't see anything controversial in these changes, LGTM per our discussions. Of course there's a lot, but it's all due to the snapd-testing-tools, as mentioned.

**Pull request merged**
